e2core/server/handlers.go

```diff
@@ -1,119 +1,131 @@
 	"github.com/suborbital/systemspec/tenant"
-	"github.com/suborbital/vektor/vk"
 )
 
-func (s *Server) executePluginByNameHandler() vk.HandlerFunc {
-	return func(w http.ResponseWriter, r *http.Request, ctx *vk.Ctx) error {
-		ident := readParam(ctx, "ident")
-		namespace := readParam(ctx, "namespace")
-		name := readParam(ctx, "name")
+func (s *Server) executePluginByNameHandler() echo.HandlerFunc {
+	return func(c echo.Context) error {
+		ident := c.Param("ident")
+		namespace := c.Param("namespace")
+		name := c.Param("name")
 
 		mod := s.syncer.GetModuleByName(ident, namespace, name)
 		if mod == nil {
-			return vk.E(http.StatusNotFound, "module not found")
+			return echo.NewHTTPError(http.StatusNotFound, "module not found")
 		}
 
-		req, err := request.FromVKRequest(r, ctx)
+		req, err := request.FromEchoContext(c)
 		if err != nil {
-			ctx.Log.Error(errors.Wrap(err, "failed to request.FromVKRequest"))
-			return vk.E(http.StatusInternalServerError, "failed to handle request")
+			return echo.NewHTTPError(http.StatusInternalServerError, "failed to handle request").SetInternal(err)
 		}
 
-		req.UseSuborbitalHeaders(r, ctx)
+		err = req.UseSuborbitalHeaders(c)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, "failed to handle request").SetInternal(err)
+		}
 
 		steps := []tenant.WorkflowStep{{FQMN: mod.FQMN}}
 
 		// a sequence executes the handler's steps and manages its state.
 		seq, err := sequence.New(steps, req)
 		if err != nil {
-			ctx.Log.Error(errors.Wrap(err, "failed to sequence.New"))
-			return vk.E(http.StatusInternalServerError, "failed to handle request")
+			return echo.NewHTTPError(http.StatusInternalServerError, "failed to handle request")
 		}
 
 		if err := s.dispatcher.Execute(seq); err != nil {
-			ctx.Log.Error(errors.Wrap(err, "failed to Execute"))
-			return vk.E(http.StatusInternalServerError, "failed to execute plugin")
+			return echo.NewHTTPError(http.StatusInternalServerError, "failed to execute plugin").SetInternal(err)
 		}
 
 		// handle any response headers that were set by the Runnables.
 		if req.RespHeaders != nil {
 			for head, val := range req.RespHeaders {
-				w.Header().Add(head, val)
+				// need to directly assign because .Add and .Set will filter out non-standard
+				// header names, which ours are.
+				if c.Response().Header()[head] == nil {
+					c.Response().Header()[head] = make([]string, 0)
+				}
+
+				c.Response().Header()[head] = append(c.Response().Header()[head], val)
 			}
 		}
 
 		responseData := seq.Request().State[mod.FQMN]
 
-		return vk.RespondBytes(ctx.Context, w, responseData, http.StatusOK)
+		return c.Blob(http.StatusOK, "text/plain", responseData)
 	}
 }
 
-func (s *Server) executePluginByRefHandler() vk.HandlerFunc {
-	return func(w http.ResponseWriter, r *http.Request, ctx *vk.Ctx) error {
-		ref := readParam(ctx, "ref")
+func (s *Server) executePluginByRefHandler(l zerolog.Logger) echo.HandlerFunc {
```

Comment:
Is there any way to attach the logger to the echo context instead of passing it in? One of the things Vektor was good at was creating a request-scoped logger for each request that included the Request UUID as a structured param, which makes debugging easier when you want to filter for all log lines associated with a specific request. 

That doesn't have to be addressed in this PR, but it's something we use regularly when debugging so I'd want to see it across all our projects.

Reply:
Technically yes, but I don't want to do that, because a logger is not request context. To log the request ID, we can do this:

```go
func (s *Server) executePluginByRefHandler(l zerolog.Logger) echo.HandlerFunc {
    return func(c *echo.Context) error {
        ll := l.With().Str("requestID", c.Request().Header().Value(echo.XRequestID).Logger()
        
        ll.Info().Msg("something happened") // this will already have the request ID in it
    }
}
```

The logger is owned by the method / handler, and logs whatever happens there. The logger is not owned by the request, if that makes sense.

Taking the logger out of the vk.Ctx was one of the biggest things I wanted to achieve.

Reply:
`ll` is short for `LocalLogger` btw that decorates the passed in logger with handler / function specific things, it's a custom I try to use on a fairly consistent basis. 🙂

Reply:
or if we don't want to create an entirely new logger, we can do this

```go
func (s *Server) executePluginByRefHandler(l zerolog.Logger) echo.HandlerFunc {
    return func(c *echo.Context) error {
        l.Info().Str("requestID", c.Request().Header().Value(echo.XRequestID)
    }
}
``

Reply:
That said, we can absolutely extend the context, however I would make it so grabbing the request ID off of it is easier so we don't need to do the header / context juggling, as opposed to storing a scoped logger in it.

Opened https://github.com/suborbital/go-kit/issues/18 to track that

Reply:
gokit 18 has a PR

Reply:
I won't block, it simply feels like having a request-scoped logger available automatically (such that no developer ever forgets to use it by accident) is a good idea.